### PR TITLE
[FIX] link_tracker: clean code and fix add a constraint

### PR DESCRIPTION
**Pull request by std-odoo**
Purpose
=======
Add a unicity constraint (main fields must have a unique combination).
As we must care about null values, we can not implement it in SQL
(possible in pure SQL, but we will need to create an index on the
columns, which is not possible in Odoo if we want to customize the error message).

Clean the code, before the "search or create" behavior was implemented
in the create, causing strange behavior (creating a link tracker with
the same values as an existing one didn't raise an error, but open
the form view of the existing link tracker).

### CODE REVIEW

File: addons/link_tracker/models/link_tracker.py [hunk 1]
```diff
@@ -80,6 +80,23 @@ def _get_title_from_url(self, url):
 
         return title
 
+    @api.model
+    def search_or_create(self, vals):
+        if 'url' not in vals:
+            raise ValueError('URL field required')
+        else:
```
Comment (fmdl) [inline addons/link_tracker/models/link_tracker.py:87]: no need else

```python
if 'url' not in vals:
     raise ValueError('URL field required')
vals['url'] = tools.validate_url(vals['url'])
```

Comment (std-odoo) [reply addons/link_tracker/models/link_tracker.py:87]: (moved to master PR and resolved)

File: addons/link_tracker/models/link_tracker.py [hunk 2]
```diff
@@ -112,26 +113,60 @@ def _get_title_from_url(self, url):
 
         return title
 
-    @api.model
-    def create(self, vals):
-        create_vals = vals.copy()
+    @api.constrains('url', 'campaign_id', 'medium_id', 'source_id')
+    def _check_unicity(self):
+        """
+        Can not be implemented in SQL because we want to care about null values.
```
Comment (tde-banana-odoo) [inline addons/link_tracker/models/link_tracker.py:119]: This is not really the method's docstring, more a technical comment :)

File: addons/link_tracker/models/link_tracker.py [hunk 3]
```diff
@@ -112,26 +113,60 @@ def _get_title_from_url(self, url):
...
+        """
+        # build a query to fetch all needed link trackers at once
+        search_query = []
+        for tracker in self:
+            search_query = ['|'] + search_query + [
```
Comment (tde-banana-odoo) [inline addons/link_tracker/models/link_tracker.py:124]: maybe we could use expression.OR and .AND ?

File: addons/link_tracker/models/link_tracker.py [hunk 4]
```diff
@@ -112,26 +113,60 @@ def _get_title_from_url(self, url):
...
+                '&', '&', '&',
+                ('url', '=', tracker.url),
+                ('campaign_id', '=', tracker.campaign_id.id),
+                ('medium_id', '=', tracker.medium_id.id),
+                ('source_id', '=', tracker.source_id.id)
+            ]
+        all_link_trackers = self.search(search_query[1:])
```
Comment (tde-banana-odoo) [inline addons/link_tracker/models/link_tracker.py:131]: Meh, this [:1] is not really clean. Normally using expression.OR it should not be necessary anymore as first OR should be done with a void list + a real list (to check).

File: addons/link_tracker/models/link_tracker.py [hunk 5]
```diff
@@ -112,26 +113,60 @@ def _get_title_from_url(self, url):
...
+
+        # check for unicity
+        for tracker in self:
+            if all_link_trackers.filtered(
+                lambda l: l.url == tracker.url
+                and l.campaign_id == tracker.campaign_id
+                and l.medium_id == tracker.medium_id
+                and l.source_id == tracker.source_id
+            ) != tracker:
+                raise UserError(_('Link Tracker values must be unique'))
```
Comment (tde-banana-odoo) [inline addons/link_tracker/models/link_tracker.py:141]: We could expand the commit message and give the failing values

File: addons/link_tracker/models/link_tracker.py [hunk 6]
```diff
@@ -112,26 +113,60 @@ def _get_title_from_url(self, url):
...
 
-        if 'url' not in create_vals:
+    @api.model
+    def search_or_create(self, vals):
```
Comment (tde-banana-odoo) [inline addons/link_tracker/models/link_tracker.py:144]: As this is not a real crud method, I would move it below other CRUDs .

File: addons/link_tracker/models/link_tracker.py [hunk 7]
```diff
@@ -112,26 +113,60 @@ def _get_title_from_url(self, url):
...
+        if 'url' not in vals:
             raise ValueError('URL field required')
```
Comment (tde-banana-odoo) [inline addons/link_tracker/models/link_tracker.py:146]: Note that we could translate message :) (and maybe improve it: "Creating a Link Tracker without URL is not possible" )

File: addons/link_tracker/models/link_tracker.py [hunk 8]
```diff
@@ -112,26 +113,60 @@ def _get_title_from_url(self, url):
...
-
         result = self.search(search_domain, limit=1)
 
         if result:
             return result
 
+        return self.create(vals)
+
+    @api.model
+    def create(self, vals):
```
Comment (tde-banana-odoo) [inline addons/link_tracker/models/link_tracker.py:162]: We could make it a create multi while we are at it

File: addons/link_tracker/models/link_tracker.py [hunk 9]
```diff
@@ -112,26 +113,60 @@ def _get_title_from_url(self, url):
...
+        create_vals = vals.copy()
+
+        if 'url' not in create_vals:
+            raise ValueError('URL field required')
```
Comment (tde-banana-odoo) [inline addons/link_tracker/models/link_tracker.py:166]: Same remark about error message

File: addons/link_tracker/models/link_tracker.py [hunk 10]
```diff
@@ -112,26 +113,60 @@ def _get_title_from_url(self, url):
...
+        else:
```
Comment (tde-banana-odoo) [inline addons/link_tracker/models/link_tracker.py:167]: else is not necessary

**Review by tde-banana-odoo — COMMENTED** [review]
Review: code LGTM, some styling remarks !

I think you will have to rebase, as we removed the slides tracker creation.

I also found this one: in website_links/tests/test_ui.py , test_01_test_ui : we use contactus to create a link tracker, maybe it could also be put into a search or create .

Cheers !

File: addons/link_tracker/models/link_tracker.py (hunk 4 above)
Comment (std-odoo) [reply addons/link_tracker/models/link_tracker.py:131]: Yes you are right, not needed with expression.OR :)

File: addons/link_tracker/models/link_tracker.py (hunk 3 above)
Comment (std-odoo) [reply addons/link_tracker/models/link_tracker.py:124]: Good idea :) With expression.OR/AND I can even use list comprehension

**Comment by tde-banana-odoo** [comment]
@robodoo r+

**Pull request closed without merging**